core/embed/extmod/modtrezorui/display.c

```diff
@@ -384,8 +389,8 @@ void display_icon(int x, int y, int w, int h, const void *data,
     int st = uzlib_uncompress(&decomp);
     if (st == TINF_DONE) break;  // all OK
     if (st < 0) break;           // error
-    const int px = pos % w;
-    const int py = pos / w;
+    const int px = (pos * 2) % w;
+    const int py = (pos * 2) / w;
```

Comment:
Why this change?

Reply:
The uncompress function returns data for 2 pixels which are fed to the display - therefore, the position on the display actually moves twice faster than the for loop.

Still not perfect probably, might be better to check the position of the pixels individually for cases where display dimensions are odd:
```
    int px = (pos * 2) % w;
    int py = (pos * 2) / w;
    if (px >= x0 && px <= x1 && py >= y0 && py <= y1) {
      PIXELDATA(colortable[decomp_out >> 4]);
    }
    px = (pos * 2 + 1) % w;
    py = (pos * 2 + 1) / w;
    if (px >= x0 && px <= x1 && py >= y0 && py <= y1) {
      PIXELDATA(colortable[decomp_out & 0x0F]);
    }
```

Reply:
Okay, I now see where the issue was. I assume we just never hit the problem, because we never rendered icon on the edge of the display.

> Still not perfect probably

We can safely assume the display dimensions will always be even (I've never seen a display in my life that has odd width). We also always render at even coordinate thanks to check line 365: `x &= ~1;  // cannot draw at odd coordinate`, so this is not a problem either.

So let's keep the change as it is. 👍

Reply:
The issue also applies to icons with odd dimensions, which is more likely to happen

Reply:
> The issue also applies to icons with odd dimensions, which is more likely to happen

I just checked and there are no icons with odd width in the repository. Trying to use `./core/tools/toif_convert.py` on a grayscale image with odd width throws the following error:

``` python
  File ".../trezor-firmware/python/src/trezorlib/toif.py", line 70, in _from_pil_grayscale
    left, right = pixels[i], pixels[i + 1]
IndexError: image index out of range
```

My intention was never to allow this anyway -> sent PR https://github.com/trezor/trezor-firmware/pull/2259 to check the width of provided file.

The odd heights are present in the repository, but this should not cause any issues if I understand everything correctly.

Reply:
Agree that we are safe as long as width is even. Lets keep the check simple then.